ARTEMIS-3915 Support PROXY protocol

jbertram opened the pull request.

jbertram commented:
FYI - the full test-suite is green on this.

gemmellr commented on docs/user-manual/proxy-protocol.adoc:
```diff
@@ -0,0 +1,33 @@
+Without PROXY protocol support the broker would see such client connections as coming from the proxy itself which can be misleading for administrators and complicate trouble-shooting.
+
+Both versions 1 & 2 of the PROXY Protocol are supported.
+Furthermore, this support is 100% transparent and requires no additional configuration.
+The broker automatically detects the use of the PROXY Protocol and manages the connection appropriately.
```
This is actually expressly forbidden by the PROXY protocol because it causes a security hole. It essentially has to be configured on so that only brokers that want to accept the PROXY protocol do, so people cant spoof their address. The protocol also requires that it be implemented such that connecting 'clients' are _required_ to use the protocol: https://github.com/haproxy/haproxy/blob/b167d545cf4b673de3c481088d7ce8ed65030106/doc/proxy-protocol.txt#L176-L182.

jbertram replied:
That makes sense.

jbertram commented on artemis-core-client/src/main/java/org/apache/activemq/artemis/spi/core/remoting/Connection.java:
```diff
@@ -135,10 +135,25 @@ default void disconnect() {
 
    /**
-    * {@return the string representation of the remote address this connection is connected to}
+    * {@return a string representation of the remote address of this connection; if this connection is made via the
+    * proxy protocol then this will be the original address, not the proxy address}
```
I see your point, but if I were to do that I think I would also have to do an `instanceof` in a lot of places since the `Connection` is what's passed around on the broker in most places. I'm open to ideas, though.

FWIW, I did remove the proxy-related methods from `RemotingConnection`.

jbertram commented on tests/e2e-tests/pom.xml:
```diff
@@ -220,6 +235,22 @@
                </execution>
+               <execution>
+                  <phase>test-compile</phase>
+                  <id>create-proxy</id>
```
Done.

jbertram committed:
ARTEMIS-3915 Support PROXY Protocol

This commit implements support the PROXY protocol so that the broker
will be able to determine a client's original IP address despite the
client's connection coming through a reverse proxy (e.g. HAProxy,
ngingx, etc.). Changes include:

 - A new Netty handler to extract relevant details out of PROXY Protocol
   messages and make them available to the broker
 - A new Netty handler to enforce the acceptor's PROXY Protocol config
 - A new chapter in the user manual
 - Updated logging to use this new data
 - Expose this new data via management
 - Disambiguate some variables names related to SOCKS proxy support
 - Embedded Netty-based PROXY server implementation for testing

brusdev commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/ProtocolHandler.java:
```diff
@@ -147,11 +156,26 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+            ctx.channel().attr(PROXY_PROTOCOL_VERSION).set(haProxyMessage.protocolVersion().toString());
+            proxyAttributesSet = true;
+            skipProxyBytes = true;
+         } else if (msg instanceof FullHttpRequest httpRequest) {
```
Should `skipProxyBytes` be false for `FullHttpRequest`?

jbertram replied:
Yes.

brusdev commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/NettyServerConnection.java:
```diff
@@ -51,4 +58,55 @@ public void setSNIHostname(String sniHostname) {
+    */
+   @Override
+   public String getRemoteAddress() {
+      String proxyProtocolSourceAddress = channel.attr(PROXY_PROTOCOL_SOURCE_ADDRESS).get();
+      String proxyProtocolSourcePort = channel.attr(PROXY_PROTOCOL_SOURCE_PORT).get();
+      if (proxyProtocolSourceAddress != null && !proxyProtocolSourceAddress.isEmpty() && proxyProtocolSourcePort != null && !proxyProtocolSourcePort.isEmpty()) {
+         return proxyProtocolSourceAddress + ":" + proxyProtocolSourcePort;
+      } else {
```
The method `getProxyAddress` could avoid code duplication, i.e.
```suggestion
      String result = getProxyAddress();
      if (result != null) {
         return result;
      } else {
```

gemmellr commented on tests/e2e-tests/src/test/java/org/apache/activemq/artemis/tests/e2e/proxy/HAProxyTest.java:
```diff
@@ -0,0 +1,253 @@
+ * You need to build the Artemis Docker image with 'mvn install -De2e-tests.skipImageBuild=false' before this test is
+ * executed.
+ */
+public class HAProxyTest extends E2ETestBase {
```
The client IP address wont be preserved for the TLS-passthrough case either unless its being done by a transparent proxy, behaving the same as a non-TLS case and getting the proxies IP instead. My understanding is that the Proxy Protocol could be used similarly in this case, with the proxy header just being inserted before the passed-through TLS handshake etc traffic rather than before some unencrypted traffic in the non-TLS case. 

That being the case it seems like something to test, especially as both the previous illegal sniffing impl and the current updated implementation seem like maybe the proxy protocol decoder is being added in the wrong place in the pipeline.

jbertram replied:
I added TLS tests.

gemmellr commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/NettyServerConnection.java:
```diff
@@ -51,4 +58,55 @@ public void setSNIHostname(String sniHostname) {
+      return channel.attr(PROXY_PROTOCOL_VERSION).get() == null ? null : channel.attr(PROXY_PROTOCOL_VERSION).get().toString();
+   }
+
+   public static String getProxyAddress(Connection connection) {
+      if (connection instanceof NettyServerConnection nettyServerConnection) {
```
Could a server-specific sub-interface be added, which the server then uses instead of the base Connection, and perhaps avoid either the need of the static method, or at least the need to instanceof here?

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/ProtocolHandler.java:
```diff
@@ -147,11 +156,26 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
       @Override
       public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
-         if (msg instanceof FullHttpRequest httpRequest) {
+         if (msg instanceof HAProxyMessage haProxyMessage) {
```
I'll look into using an independent handler for this, but I'm not sure it matters all that much because as soon as the protocol is detected this handler is removed. Therefore, as far as I can tell this boolean is only checked once per connection.

gemmellr replied:
Its only effective once per connection but it was being both checked and set for every read, including when the proxyProtocol isnt even being enabled. Now its just being checked for every read.

jbertram replied:
Done.

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/ProtocolHandler.java:
```diff
@@ -184,7 +208,9 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
-            super.channelRead(ctx, msg);
+            // slice off the proxy-related bytes that have already been read so other protocol handlers don't choke on them
+            super.channelRead(ctx, skipProxyBytes ? ((ByteBuf) msg).slice() : msg);
+            skipProxyBytes = false;
```
I refactored the logic into a customer handler. 

FWIW, AFAICT this was actually required in order to get the SSL stuff to work.

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/NettyAcceptor.java:
```diff
@@ -145,6 +146,8 @@ public class NettyAcceptor extends AbstractAcceptor {
 
    private final boolean sslEnabled;
 
+   private final boolean proxyEnabled;
```
Done.

gemmellr commented:
I havent had a chance to look properly at this after all the changes yet...but skimming the tests, I only see spotted positive tests where it was expected to work and did. We should have negative tests for this, verifying connecting doesnt work when proxyProtocolEnabled=true and a connection connects without the header and vice versa fails when proxyProtocolEnabled=false (/defaulted) a connection with the proxy-protocol header comes in.

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/HAProxyMessageHandler.java:
```diff
@@ -0,0 +1,50 @@
+import static org.apache.activemq.artemis.utils.ProxyProtocolUtil.PROXY_PROTOCOL_SOURCE_ADDRESS;
+import static org.apache.activemq.artemis.utils.ProxyProtocolUtil.PROXY_PROTOCOL_VERSION;
+
+public class HAProxyMessageHandler extends ChannelInboundHandlerAdapter {
```
I pushed a change here to simplify the handler based on reliable assumptions.

jbertram commented on docs/user-manual/proxy-protocol.adoc:
```diff
@@ -0,0 +1,59 @@
+:idseparator: -
+:docinfo: shared
+
+As noted in the official https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt[PROXY Protocol documentation]:
```
Done.

jbertram commented on tests/e2e-tests/src/test/java/org/apache/activemq/artemis/tests/e2e/proxy_protocol/HAProxyTest.java:
```diff
@@ -0,0 +1,334 @@
+      });
+      Throwable rootCause = ExceptionUtils.getRootCause(e);
+      assertNotNull(rootCause);
+      assertTrue(rootCause instanceof ActiveMQConnectionTimedOutException);
```
That sounds great, @tabish121.

In the meantime, I've added a new handler to enforce the acceptor's proxy-related config. Any violation causes the connection to be closed immediately. I also updated the `HAProxyMessageHandler` to verify expectations.

tabish121 replied:
I've create a mock HAProxy server and done some basic testing on it against this branch and it seems to work as needed.  You could have a go at porting your tests to using that and see if we need to make any more adjustments to it. Branch is linked below:

https://github.com/tabish121/activemq-artemis/tree/mock-haproxy-server

jbertram replied:
I'm converting this to draft while I work on transitioning the tests from Testcontainers to the embedded HAProxy from Tim.

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/HAProxyMessageEnforcer.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class HAProxyMessageEnforcer extends ByteToMessageDecoder {
+
+   private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
Fixed.

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/NettyAcceptor.java:
```diff
@@ -472,6 +482,14 @@ public synchronized void start() throws Exception {
          @Override
          public void initChannel(Channel channel) throws Exception {
             ChannelPipeline pipeline = channel.pipeline();
+            if (proxyProtocolEnabled) {
```
Not sure how I missed that. :laughing:

jbertram commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/impl/netty/HAProxyMessageEnforcer.java:
```diff
@@ -0,0 +1,73 @@
+   @Override
+   public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+      ByteBuf in = (ByteBuf) msg;
+      if (in.readableBytes() < 4) {
```
Check out the latest push. I added your test, and I removed the enforcer object for the normal (i.e. `proxyProtocolEnabled=false`) case although I still check in the `ProtocolDecoder` so we can fail fast.

jbertram commented on tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/proxyprotocol/HAProxyTest.java:
```diff
@@ -0,0 +1,411 @@
+
+   @Test
+   @Timeout(30)
+   public void testBrokerHandlesOutOfOrderDeliveryIdInTransfer() throws Exception {
```
I didn't even look at the name. :laughing: 

Fixed now with your suggestion.
